pyigm/metallicity/pdf.py

```diff
@@ -0,0 +1,154 @@
+""" Class for Metallicity PDF
+"""
+from __future__ import print_function, absolute_import, division, unicode_literals
+
+import numpy as np
+import pdb
+
+
+class MetallicityPDF(object):
+    """A Class for Metallicity PDF
+
+    Attributes
+    ----------
+    ZH : ndarray
+      Values where the PDF has been evaluated
+    phi_ZH : ndarray
+      PDF of the metallicity
+    """
+
+    def __init__(self, ZH, pdf_ZH, parent=None, dZH=None):
+        """
+        Parameters
+        ----------
+        ZH : ndarray
+          Values of ZH (log10, typically but not required) where the PDF has been evaluated
+        pdf_ZH : ndarray
+          PDF of the metallicity
+        parent : object, optional
+          Parent of the metalliicty PDF, e.g. an AbsSystem
+        dZH : float or ndarray
+          Width of the ZH 'bins' (log 10)
+
+        Returns
+        -------
+        MetallicityPDF
+
+        """
+        # Tests
+        if ZH.size != pdf_ZH.size:
+            raise IOError("ZH and pdf_ZH must have the same size")
+        # Set
+        self.ZH = ZH
+        self.pdf_ZH = pdf_ZH
+        self.parent = parent
+
+        # dZH
+        if dZH is None:
+            dZH = self.ZH - np.roll(self.ZH,1)
+            dZH[0] = dZH[1]
+            self.dZH = dZH
+        else:
+            self.dZH = dZH
+
+        # Normalize pdf_ZH
+        self.normalize()
+
+    @property
+    def meanZH(self):
+        """ Calculate and return average ZH from the PDF
+
+        Weighted in log-space
+
+        Returns
+        -------
+        meanZH : float
+        """
+        meanZH = np.sum(self.ZH*self.pdf_ZH*self.dZH)
+        return meanZH
+
+    def confidence_limits(self, cl):
+        """ Calculate the bounds of a given confidence interval
+
+        Spline interpolation is used
+
+        Parameters
+        ----------
+        cl : float
+          Confidence interval, ranging from 0 to 1
+
+        Returns
+        -------
+        ZH_min, ZH_max : float, float
+          Bounds corresponding to the input confidence limit
+        """
+        from scipy.interpolate import interp1d
+
+        if (cl <= 0.) or (cl >=1):
+            raise IOError("cl must range from 0-1")
+        # Spline the PDF cumulative sum vs ZH
+        cumul = np.cumsum(self.dZH*self.pdf_ZH)
+        f = interp1d(cumul, self.ZH)
+        # Caculate
+        frac = (1.- cl) / 2.
+        ZH_min = float(f(frac))
+        ZH_max = float(f(1-frac))
+        # Return
+        return ZH_min, ZH_max
+
+    def normalize(self):
+        """ Normalize the PDF
+
+        First checks to see if it is normalized
```

Comment:
I was going to check and then decided not to (but left
the comment in).  I think it best not to check but just do.